pkg/maps/ctmap/per_cluster_ctmap.go

```diff
@@ -102,379 +123,124 @@ type PerClusterCTMapVal struct {
-		gm.any6 = make(map[uint32]struct{})
-	}
-
-	return gm
-}
-
-func (gm *dummyPerClusterCTMaps) UpdateClusterCTMaps(clusterID uint32) error {
+func (gm *perClusterCTMaps) DeleteClusterCTMaps(clusterID uint32) error {
 	if err := cmtypes.ValidateClusterID(clusterID); err != nil {
 		return err
 	}
 
 	gm.Lock()
 	defer gm.Unlock()
 
-	if gm.ipv4 {
-		gm.tcp4[clusterID] = struct{}{}
-		gm.any4[clusterID] = struct{}{}
-	}
-
-	if gm.ipv6 {
-		gm.tcp6[clusterID] = struct{}{}
-		gm.any6[clusterID] = struct{}{}
-	}
-
-	return nil
+	gm.clusterIDs.Delete(clusterID)
+	return gm.foreach(
+		func(om *PerClusterCTMap) error { return om.deleteClusterCTMap(clusterID) },
+	)
 }
 
-func (gm *dummyPerClusterCTMaps) DeleteClusterCTMaps(clusterID uint32) error {
-	if err := cmtypes.ValidateClusterID(clusterID); err != nil {
-		return err
-	}
-
+func (gm *perClusterCTMaps) GetAllClusterCTMaps() []*Map {
 	gm.Lock()
 	defer gm.Unlock()
 
-	if gm.ipv4 {
-		delete(gm.tcp4, clusterID)
-		delete(gm.any4, clusterID)
-	}
-
-	if gm.ipv6 {
-		delete(gm.tcp6, clusterID)
-		delete(gm.any6, clusterID)
+	var maps []*Map
+	for clusterID := range gm.clusterIDs {
+		gm.foreach(func(om *PerClusterCTMap) error {
+			maps = append(maps, om.newInnerMap(clusterID))
+			return nil
+		})
 	}
-
-	return nil
+	return maps
 }
 
-func (gm *dummyPerClusterCTMaps) GetClusterCTMaps(clusterID uint32) []*Map {
-	ims := []*Map{}
-
+func (gm *perClusterCTMaps) getClusterCTMaps(clusterID uint32) ([]*Map, error) {
 	if err := cmtypes.ValidateClusterID(clusterID); err != nil {
-		return []*Map{}
+		return nil, err
 	}
 
 	gm.Lock()
 	defer gm.Unlock()
 
-	if gm.ipv4 {
-		if _, ok := gm.tcp4[clusterID]; ok {
-			ims = append(ims, &Map{})
-		}
-		if _, ok := gm.any4[clusterID]; ok {
-			ims = append(ims, &Map{})
-		}
-	}
+	var maps []*Map
+	gm.foreach(func(om *PerClusterCTMap) error {
+		maps = append(maps, om.newInnerMap(clusterID))
+		return nil
+	})
 
-	if gm.ipv6 {
-		if _, ok := gm.tcp6[clusterID]; ok {
-			ims = append(ims, &Map{})
-		}
-		if _, ok := gm.any6[clusterID]; ok {
-			ims = append(ims, &Map{})
-		}
-	}
-
-	return ims
+	return maps, nil
 }
 
-func (gm *dummyPerClusterCTMaps) GetAllClusterCTMaps() ([]*Map, error) {
-	ims := []*Map{}
-
-	gm.RLock()
-	defer gm.RUnlock()
-
-	for i := uint32(1); i <= cmtypes.ClusterIDMax; i++ {
-		if gm.ipv4 {
-			if _, ok := gm.tcp4[i]; ok {
-				ims = append(ims, &Map{})
-			}
-			if _, ok := gm.any4[i]; ok {
-				ims = append(ims, &Map{})
-			}
-		}
-		if gm.ipv6 {
-			if _, ok := gm.tcp6[i]; ok {
-				ims = append(ims, &Map{})
-			}
-			if _, ok := gm.any6[i]; ok {
-				ims = append(ims, &Map{})
-			}
-		}
-	}
+func (gm *perClusterCTMaps) cleanup() error {
+	gm.Lock()
+	defer gm.Unlock()
 
-	return ims, nil
+	return gm.foreach(func(om *PerClusterCTMap) error {
+		return om.cleanup()
+	})
 }
 
-func (gm *dummyPerClusterCTMaps) Cleanup() {
-	gm.RLock()
-	defer gm.RUnlock()
+func (gm *perClusterCTMaps) foreach(fn func(om *PerClusterCTMap) error) error {
```

Comment:
I do not fully understand what happens when the operation against one of the maps is failed. Should we rollback the changes or do you think is it ok to leave it as is?

Reply:
Honestly this is also the aspect I was also doubtful about. At the end, I currently adopted the simplest approach, always performing the given operation on all maps, without attempting any rollback process on the already processed maps in case of errors. 

This is because (a) it is likely that the consumer of these methods will retry again the same operation (e.g., through a controller) in case an error has occurred, (b) errors seem unlikely (as syscall errors), but if they happen we probably ended up in a broken scenario and also the subsequent rollback may also fail. 

All in all, my feeling is that implementing the rollback would make the overall logic more complex and fragile (as it is difficult to actually test the error cases), with only very limited benefit. Additionally, it would also be unclear how to rollback deletion operations, as recreating the map would probably not be the best option. 

As for the "process all and concatenate the errors" vs "abort on first error" approaches, I personally preferred the first because it works better especially in case of deletion/cleanup operations, ensuring for instance that we attempt to delete all maps even if one of the earlier operations failed. This is especially useful as in certain cases we might not want to subsequently retry failed deletion operations.

I'm obviously open to reconsider the above if you feel that roll backing the changes on errors would be beneficial.

Reply:
Yep, all makes sense to me. Thanks for your explanation. I have no objection to going with the current direction 👍